plugins/sigma/src/plugin.cpp

```diff
@@ -0,0 +1,211 @@
+        return std::move(yaml.error());
+      }
+      auto rule = parse_rule(*yaml);
+      if (not rule) {
+        return std::move(rule.error());
+      }
+      rules[path.string()] = {std::move(*yaml), std::move(*rule)};
+      return {};
+    }
+
+    std::filesystem::path path;
+    std::unordered_map<std::string, std::pair<data, expression>> rules = {};
+  };
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    auto state = monitor_state{};
+    state.path = path_;
+    if (auto ok = state.update(state.path); not ok) {
+      diagnostic::error("failed to load sigma rules: {}", ok.error())
+        .emit(ctrl.diagnostics());
+    }
+    auto last_update = std::chrono::steady_clock::now();
+    co_yield {}; // signal that we're done initializing
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      if (last_update + refresh_interval_ < std::chrono::steady_clock::now()) {
+        if (auto ok = state.update(state.path); not ok) {
+          diagnostic::warning("failed to update sigma rules: {}", ok.error())
+            .emit(ctrl.diagnostics());
+        }
+        last_update = std::chrono::steady_clock::now();
+      }
+      for (const auto& [path, entry] : state.rules) {
+        const auto& [yaml, rule] = entry;
+        auto expr = tailor(rule, slice.schema());
+        if (not expr) {
+          continue;
+        }
+        if (auto event = filter(slice, *expr)) {
+          const auto rule_schema = caf::get<record_type>(type::infer(yaml));
+          const auto result_schema = type{
+            "tenzir.sigma",
+            record_type{
+              {"event", event->schema()},
+              {"rule", rule_schema},
+            },
+          };
+          auto result_builder
+            = result_schema.make_arrow_builder(arrow::default_memory_pool());
+          auto array = to_record_batch(*event)->ToStructArray().ValueOrDie();
+          for (const auto& row :
+               values(caf::get<record_type>(slice.schema()), *array)) {
+            const auto append_row_result
+              = caf::get<arrow::StructBuilder>(*result_builder).Append();
+            TENZIR_ASSERT(append_row_result.ok());
+            const auto append_event_result = append_builder(
+              caf::get<record_type>(event->schema()),
+              caf::get<arrow::StructBuilder>(
+                *caf::get<arrow::StructBuilder>(*result_builder)
+                   .field_builder(0)),
+              *row);
+            TENZIR_ASSERT(append_event_result.ok());
+            const auto append_rule_result = append_builder(
+              rule_schema,
+              caf::get<arrow::StructBuilder>(
+                *caf::get<arrow::StructBuilder>(*result_builder)
+                   .field_builder(1)),
+              caf::get<view<record>>(make_view(yaml)));
+            TENZIR_ASSERT(append_rule_result.ok());
```

Comment:
I noticed that Sigma rule hits attach the entire rule to every event. This attachment seems to copy all information from the Sigma rule file, which includes potentially useless and potentially intimate info like the description and author.

1. What was the reasoning for this again?
2. Can they be grouped in a later version of the operator? Maybe just under the rule title?

Reply:
You typically use Sigma rules to run detections. The information which Sigma rule matched is critical here. It's not uncommon to have hundreds of rules, and sometimes you may even create multiple results for a single event if multiple rules match.

On (2), I think the idea would be to put the Sigma functionality somewhere into the matcher, once we've made that pipeline-native. We have that on the roadmap already and it's coming up. That's probably the natural evolution of this operator that currently serves as a good proof of concept.